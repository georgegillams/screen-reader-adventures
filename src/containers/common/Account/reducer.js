import produce from 'immer';

import { logout, requestVerificationEmail } from './actions';

export const initialState = {
  loggingOut: false,
  logOutError: null,
  logOutResult: null,
  requestingVerificationEmail: false,
  requestVerificationEmailError: null,
  requestVerificationEmailResult: null,
};

const reducer = (state = initialState, { type, payload }) =>
  produce(state, draft => {
    switch (type) {
      case logout.REQUEST:
        draft.loggingOut = true;
        draft.logOutError = null;
        break;

      case logout.SUCCESS:
        draft.loggingOut = false;
        draft.logOutResult = payload;
        break;

      case logout.FAILURE:
        draft.loggingOut = false;
        draft.logOutError = payload;
        break;

      case requestVerificationEmail.REQUEST:
        draft.requestingVerificationEmail = true;
        draft.requestVerificationEmailError = null;
        break;

      case requestVerificationEmail.SUCCESS:
        draft.requestingVerificationEmail = false;
        draft.requestVerificationEmailResult = payload;
        break;

      case requestVerificationEmail.FAILURE:
        draft.requestingVerificationEmail = false;
        draft.requestVerificationEmailError = payload;
        break;
    }
  });

export default reducer;
