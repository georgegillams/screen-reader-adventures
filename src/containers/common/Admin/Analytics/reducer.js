import produce from 'immer';

import { load } from './actions';

export const initialState = {
  analytics: null,
  loading: false,
  loadError: null,
};

const reducer = (state = initialState, { type, payload }) =>
  produce(state, draft => {
    switch (type) {
      case load.REQUEST:
        draft.loading = true;
        draft.loadError = null;
        break;

      case load.SUCCESS:
        draft.loading = false;
        draft.analytics = payload;
        break;

      case load.FAILURE:
        draft.loading = false;
        draft.loadError = payload;
        break;
    }
  });

export default reducer;
