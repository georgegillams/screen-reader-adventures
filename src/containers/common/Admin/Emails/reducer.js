import produce from 'immer';

import { load, resend } from './actions';

export const initialState = {
  emails: null,
  loading: false,
  loadError: null,

  emailToResend: null,
  resending: false,
  resendError: null,
};

const reducer = (state = initialState, { type, payload }) =>
  produce(state, draft => {
    switch (type) {
      case load.REQUEST:
        draft.loading = true;
        draft.loadError = null;
        break;

      case load.SUCCESS:
        draft.loading = false;
        draft.emails = payload;
        break;

      case load.FAILURE:
        draft.loading = false;
        draft.loadError = payload;
        break;

      case resend.TRIGGER:
        draft.emailToResend = payload.emailToResend;
        break;

      case resend.REQUEST:
        draft.resending = true;
        draft.resendError = null;
        break;

      case resend.SUCCESS:
        draft.resending = false;
        break;

      case resend.FAILURE:
        draft.resending = false;
        draft.resendError = payload;
        break;
    }
  });

export default reducer;
