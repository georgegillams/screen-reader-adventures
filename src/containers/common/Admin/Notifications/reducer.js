import produce from 'immer';

import { load, create, remove } from './actions';

export const initialState = {
  notifications: null,
  loading: false,
  loadError: null,

  notificationToCreate: null,
  creating: false,
  createError: null,

  notificationToRemove: null,
  removing: false,
  removeError: null,
};

const reducer = (state = initialState, { type, payload }) =>
  produce(state, draft => {
    switch (type) {
      case load.REQUEST:
        draft.loading = true;
        draft.loadError = null;
        break;

      case load.SUCCESS:
        draft.loading = false;
        draft.notifications = payload;
        break;

      case load.FAILURE:
        draft.loading = false;
        draft.loadError = payload;
        break;

      case create.TRIGGER:
        draft.notificationToCreate = payload;
        break;

      case create.REQUEST:
        draft.creating = true;
        draft.createError = null;
        break;

      case create.SUCCESS:
        draft.creating = false;
        break;

      case create.FAILURE:
        draft.creating = false;
        draft.createError = payload;
        break;

      case remove.TRIGGER:
        draft.notificationToRemove = payload;
        break;

      case remove.REQUEST:
        draft.removing = true;
        draft.removeError = null;
        break;

      case remove.SUCCESS:
        draft.removing = false;
        break;

      case remove.FAILURE:
        draft.removing = false;
        draft.removeError = payload;
        break;
    }
  });

export default reducer;
