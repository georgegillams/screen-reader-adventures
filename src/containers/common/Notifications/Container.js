import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import DebugObject from 'components/common/DebugObject';
import Notification from 'gg-components/Notification/Notification';
import STYLES from './notifications.scss';
import { cssModules } from 'gg-components/helpers/cssModules';
import useTabMadeVisible from 'client-utils/common/useTabMadeVisible';

const getClassName = cssModules(STYLES);

const Notifications = props => {
  const { load, notificationsState } = props;
  const { notifications } = notificationsState;

  useTabMadeVisible(load);
  useEffect(() => {
    load();
  }, []);

  const filteredNotifications =
    notifications &&
    notifications.filter(n => {
      if (n.deleted) {
        return false;
      }

      if (!n.type || !n.message) {
        return false;
      }

      return true;
    });

  return (
    <>
      {filteredNotifications &&
        filteredNotifications.map(n => (
          <Notification key={n.id} className={getClassName('notifications__notification')} type={n.type}>
            {n.message}
          </Notification>
        ))}
      <div>
        <DebugObject debugTitle="Notifications" debugObject={{ load, notificationsState }} />
      </div>
    </>
  );
};

Notifications.propTypes = {
  load: PropTypes.func.isRequired,
  notificationsState: PropTypes.shape({
    loading: PropTypes.bool,
    loadError: PropTypes.object,
    notifications: PropTypes.arrayOf(PropTypes.object),
  }).isRequired,
};

export default Notifications;
