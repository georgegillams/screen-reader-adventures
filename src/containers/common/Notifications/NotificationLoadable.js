import Loadable from 'react-loadable';

const Loading = () => null;

const NotificationLoadable = Loadable({
  loader: () => import('gg-components/Notification/Notification.js'),
  loading: Loading,
});

export default NotificationLoadable;
