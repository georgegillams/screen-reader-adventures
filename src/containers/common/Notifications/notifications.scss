@import '~utils/tokens';

.notifications {
  &__notification {
    &:not(:last-of-type) {
      margin-bottom: $spacing-base;
    }
  }
}
