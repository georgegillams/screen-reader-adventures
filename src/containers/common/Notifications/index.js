import { memo } from 'react';

import { compose } from 'redux';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';

import { load } from 'containers/common/Admin/Notifications/actions';
import { selectState } from 'containers/common/Admin/Notifications/selectors';
import Notifications from './Container';
import injectSaga from 'client-utils/common/redux/inject-saga';
import injectReducer from 'client-utils/common/redux/inject-reducer';

import { KEY } from 'containers/common/Admin/Notifications/constants';
import saga from 'containers/common/Admin/Notifications/saga';
import reducer from 'containers/common/Admin/Notifications/reducer';

const mapStateToProps = createStructuredSelector({
  notificationsState: selectState(),
});

export function mapDispatchToProps(dispatch) {
  return {
    load: payload => dispatch(load(payload)),
  };
}

const withConnect = connect(mapStateToProps, mapDispatchToProps);
const withSaga = injectSaga({ key: KEY, saga });
const withReducer = injectReducer({ key: KEY, reducer });

export default compose(withSaga, withReducer, withConnect, memo)(Notifications);
