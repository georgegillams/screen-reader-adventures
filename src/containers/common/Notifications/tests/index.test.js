import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { initialState } from 'containers/common/Admin/Notifications/reducer';

import configureStore from 'client-utils/common/redux/configure-store';

import NotificationsIndex from '../index';
import Notifications from '../Container';

describe('<Notifications />', () => {
  let store;
  const spy = jest.fn();

  beforeAll(() => {
    store = configureStore({});
  });

  it('should render correctly - index', () => {
    const { container } = render(
      <Provider store={store}>
        <NotificationsIndex />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly with initial state', () => {
    const { container } = render(
      <Provider store={store}>
        <Notifications
          load={spy}
          notificationsState={{
            ...initialState,
          }}
        />
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });

  it('should render correctly with user', () => {
    const { container } = render(
      <Provider store={store}>
        <Notifications
          load={spy}
          notificationsState={{
            ...initialState,
          }}
        />
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });

  it('should render correctly with loading=true', () => {
    const { container } = render(
      <Provider store={store}>
        <Notifications
          load={spy}
          notificationsState={{
            ...initialState,
            loading: true,
          }}
        />
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });

  it('should render correctly with loadError', () => {
    const { container } = render(
      <Provider store={store}>
        <Notifications
          load={spy}
          notificationsState={{
            ...initialState,
            loadError: { error: 'some_error', errorMessage: 'Something went wrong' },
          }}
        />
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });

  it('should render correctly with notifications', () => {
    const { container } = render(
      <Provider store={store}>
        <Notifications
          load={spy}
          notificationsState={{
            ...initialState,
            notifications: [
              { id: '1', message: 'message1', type: 'success' },
              { id: '2', message: 'message2', type: 'warn' },
              { id: '3', message: 'message3', type: 'error' },
            ],
          }}
        />
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });
});
