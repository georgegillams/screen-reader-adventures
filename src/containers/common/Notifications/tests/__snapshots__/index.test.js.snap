// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Notifications /> should render correctly - index 1`] = `
<div>
  <div />
</div>
`;

exports[`<Notifications /> should render correctly with initial state 1`] = `
<div>
  <div />
</div>
`;

exports[`<Notifications /> should render correctly with loadError 1`] = `
<div>
  <div />
</div>
`;

exports[`<Notifications /> should render correctly with loading=true 1`] = `
<div>
  <div />
</div>
`;

exports[`<Notifications /> should render correctly with notifications 1`] = `
<div>
  <div
    class="notification-comp__notification notification-comp__notification--success notifications__notification"
  >
    message1
  </div>
  <div
    class="notification-comp__notification notification-comp__notification--warn notifications__notification"
  >
    message2
  </div>
  <div
    class="notification-comp__notification notification-comp__notification--error notifications__notification"
  >
    message3
  </div>
  <div />
</div>
`;

exports[`<Notifications /> should render correctly with user 1`] = `
<div>
  <div />
</div>
`;
